git/mktag.go

```diff
@@ -0,0 +1,145 @@
+package git
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+// Implements mktag by reading the input from r in the format
+// described in git-mktag(1).
+func Mktag(c *Client, r io.Reader) (Sha1, error) {
+	// This goes through a ridiculous amount of effort to ensure, not only that we
+	// error out in the same conditions as the official git client, but also that we
+	// error out with the exact same message (and character position), even when multiple
+	// things are wrong or the error message is misleading. The payoff is that we can run
+	// the official git test suite against it.
+	//
```

Comment:
Yes, this is the double edged sword for both the goal of being a drop-in replacement for official git and also to use the official git test suite. :(